### PR TITLE
AN-260 Rewrite

### PR DESCRIPTION
**Pull request by USMCFighter94**
Kotlin all the things!

### CODE REVIEW

File: bubble-rank-view/src/main/java/com/stockx/bubblerank/BubbleRank.kt [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+    @SuppressLint("SwitchIntDef")
+    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
+        val widthSize = MeasureSpec.getSize(widthMeasureSpec)
+        val heightMode = MeasureSpec.getMode(heightMeasureSpec)
+        val heightSize = MeasureSpec.getSize(heightMeasureSpec)
+
+        val scalar = circleCount - 1
+        val numerator = (widthSize + (2 * offset * scalar)).toFloat()
+        val denominator = 2 * BIG_RAD_SCALE + scalar * 2
```
Comment (philboyd) [inline bubble-rank-view/src/main/java/com/stockx/bubblerank/BubbleRank.kt]: 2 is used a lot in this file, but I'm not sure what the purpose of it is. Are you able to move it a variable?

File: bubble-rank-view/src/main/java/com/stockx/bubblerank/BubbleRank.kt [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+                    Matrix.ScaleToFit.END)
+        }
+        return Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
+    }
+
+    private fun drawCirclesAndText(index: Int, canvas: Canvas?) {
+        val current = index == currentRank
+        var centerX = index * 2 * smallRadius + smallRadius - index * offset * 2
+
+        centerX += if (current) bigRadius - smallRadius else 2 * (bigRadius - smallRadius)
```
Comment (philboyd) [inline bubble-rank-view/src/main/java/com/stockx/bubblerank/BubbleRank.kt]: May want to break this up to multiline. Helps create a clear separation of condition and statement

File: bubble-rank-view/src/main/java/com/stockx/bubblerank/BubbleRank.kt [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+                it.setLocalMatrix(Matrix().apply {
+                    setTranslate((centerX - (if (current) bigRadius else smallRadius)).toFloat(), if (current) 0f else bigRadius.toFloat())
+                })
+                circlePaint.setShader(it)
+            }
+        } else {
+            circlePaint.shader = null
+        }
+
+        val radius = (if (current) bigRadius else smallRadius).toFloat()
```
Comment (philboyd) [inline bubble-rank-view/src/main/java/com/stockx/bubblerank/BubbleRank.kt]: You could move this up so you can use `radius` on line 135.

File: bubble-rank-example/build.gradle [hunk 4]
```diff
@@ -1,36 +1,21 @@
 apply plugin: 'com.android.application'
+apply plugin: 'kotlin-android'
 
 android {
     compileSdkVersion 28
-    buildToolsVersion '28.0.3'
 
     defaultConfig {
         applicationId "com.example.bubblerank"
```
Comment (philboyd) [inline bubble-rank-example/build.gradle]: lol

File: bubble-rank-view/src/main/java/com/stockx/bubblerank/BubbleRank.kt [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+        canvas?.apply {
+            drawCircle(centerX.toFloat(), centerY.toFloat(), radius, circlePaint)
+            drawCircle(centerX.toFloat(), centerY.toFloat(), radius, circleBorderPaint)
+        }
+
+        if (!current) {
+            textPaint.apply {
+                textSize = resources.getDimension(R.dimen.rank_default_text_size)
+                getTextBounds("4", 0, 1, bounds)
```
Comment (torresmi) [inline bubble-rank-view/src/main/java/com/stockx/bubblerank/BubbleRank.kt]: What is the "4" for?

File: bubble-rank-view/src/main/java/com/stockx/bubblerank/BubbleRank.kt (hunk 1 above)
Comment (USMCFighter94) [reply bubble-rank-view/src/main/java/com/stockx/bubblerank/BubbleRank.kt]: Converting radius to diameter

File: bubble-rank-view/src/main/java/com/stockx/bubblerank/BubbleRank.kt (hunk 5 above)
Comment (USMCFighter94) [reply bubble-rank-view/src/main/java/com/stockx/bubblerank/BubbleRank.kt]: Your guess is as good as mine

Comment (torresmi) [reply bubble-rank-view/src/main/java/com/stockx/bubblerank/BubbleRank.kt]: 😞 I'm guessing it's an arbitrary number to get the size of a single digit number.

**Pull request merged**